File: src/context/klawed_md.c
/*
 * klawed_md.c - Read and parse KLAWED.md file
 */

#include <stdio.h>
#include <stdlib.h>
#include <sys/stat.h>
#include <limits.h>

#include "klawed_md.h"

/**
 * Read KLAWED.md from the working directory if it exists.
 * Reads the entire file into memory.
 */
char* read_klawed_md(const char *working_dir) {
    if (!working_dir) {
        return NULL;
    }

    char klawed_md_path[PATH_MAX];
    snprintf(klawed_md_path, sizeof(klawed_md_path), "%s/KLAWED.md", working_dir);

    // Check if file exists
    struct stat st;
    if (stat(klawed_md_path, &st) != 0) {
        return NULL; // File doesn't exist
    }

    // Read the file
    FILE *f = fopen(klawed_md_path, "r");
    if (!f) {
        return NULL;
    }

    // Allocate buffer based on file size
    size_t file_size = (size_t)st.st_size;
    char *content = malloc(file_size + 1);
    if (!content) {
        fclose(f);
        return NULL;
    }

    size_t read_size = fread(content, 1, file_size, f);
    fclose(f);

    if (read_size != file_size) {
        free(content);
        return NULL;
    }

    content[file_size] = '\0';
    return content;
}

File: zig/context/klawed_md.zig
//! context/klawed_md.zig — Read KLAWED.md from the project directory
//!
//! Zig port of src/context/klawed_md.c.
//!
//! Reads the `KLAWED.md` file from the given working directory and returns
//! its contents as an owned string.  Returns `null` if the file does not exist.

const std = @import("std");

/// Read `KLAWED.md` from `working_dir`.
/// Returns an owned string (caller must free), or `null` if the file does not exist.
pub fn readKlawedMd(allocator: std.mem.Allocator, working_dir: []const u8) !?[]u8 {
    var path_buf: [std.fs.max_path_bytes]u8 = undefined;
    const path = std.fmt.bufPrint(&path_buf, "{s}/KLAWED.md", .{working_dir}) catch
        return null;

    const file = std.fs.openFileAbsolute(path, .{}) catch |err| switch (err) {
        error.FileNotFound => return null,
        else => return err,
    };
    defer file.close();

    const stat = try file.stat();
    const content = try allocator.alloc(u8, stat.size);
    errdefer allocator.free(content);

    const n = try file.readAll(content);
    if (n != stat.size) {
        // Partial read — return what we got
        const trimmed = try allocator.realloc(content, n);
        return trimmed;
    }
    return content;
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

test "readKlawedMd returns null for non-existent directory" {
    const allocator = std.testing.allocator;
    const result = try readKlawedMd(allocator, "/nonexistent_dir_abc123");
    try std.testing.expect(result == null);
}

test "readKlawedMd reads existing KLAWED.md" {
    const allocator = std.testing.allocator;

    // Create a temporary directory with a KLAWED.md
    var tmp = std.testing.tmpDir(.{});
    defer tmp.cleanup();

    const content_str = "# KLAWED.md\n\nTest content.\n";
    const file = try tmp.dir.createFile("KLAWED.md", .{});
    defer file.close();
    try file.writeAll(content_str);

    // Get absolute path of the temp dir
    var path_buf: [std.fs.max_path_bytes]u8 = undefined;
    const tmp_path = try tmp.dir.realpath(".", &path_buf);

    const result = try readKlawedMd(allocator, tmp_path);
    defer if (result) |r| allocator.free(r);

    try std.testing.expect(result != null);
    try std.testing.expectEqualStrings(content_str, result.?);
}

test "readKlawedMd returns null when no KLAWED.md present" {
    const allocator = std.testing.allocator;

    var tmp = std.testing.tmpDir(.{});
    defer tmp.cleanup();

    var path_buf: [std.fs.max_path_bytes]u8 = undefined;
    const tmp_path = try tmp.dir.realpath(".", &path_buf);

    const result = try readKlawedMd(allocator, tmp_path);
    try std.testing.expect(result == null);
}
